### ISSUE TITLE
Allow endpoints to support promises rather than urls

### ISSUE THREAD

**Issue by joeenzminger** [FineUploader/fine-uploader#1320]
Struggling to implement CSRF on calls from the uploader in S3 mode.  Even using a customHeader function for the signature and uploadSuccess endpoints, I see occasionally failures that I haven't been able to trace down that I think happen when our server rejects signature requests due to CSRF mismatches.

Add support for these endpoints to be functions that return promises (similar to how objectProperties.key currently functions).  This would let us handle communication between the client and our server rather than delegating it to fineuploader.

**Comment by joeenzminger** [comment]
Ray - just bumping this to see if you guys are still considering making the endpoint properties accept functions that return promises rather than just strings (so we can implement our own xhr, etc rather than having fineuploader do it internally).

**Comment by rnicholus** [comment]
I would expect the promise to return a URL to the endpoint. Is this what you are suggesting?

**Comment by joeenzminger** [comment]
No.  The promise would resolve to the data that you'd expect to currently be returned from the endpoint.  This is a somewhat contrived example, but something along these lines *should* work:

```javascript
uploadSuccess: {
        endpoint: function (key, uuid, name, bucket) {
           var p = new qq.Promise();
           setTimeout(function() {
              p.success({});
           }, 1000);
           return p;
        }
    }
```

This has the advantage of allowing us to have a bit more flexibility into how our API is structured - for instance (and again contrived), if we wanted to use a Websocket in the above example rather than xhr we could do it.

I'm not talking about endpoints that talk to services outside our control (like the S3 endpoints - obviously you have to manage that internal to the uploader).  I'm talking about the endpoints you specify that are integration points with our application (uploadSuccess, getKey, signature, etc.).  Promises would give us more flexibility in implementation.

**Comment by rnicholus** [comment]
Ah I see. I think I misunderstood the intent of this case, looking back at my earlier comments. 

I can see implementing this for everything *other than* upload requests (for all endpoint types, not just S3). There would need to be some agreement as to what the resolved promise would "return". Should it return an `XMLHttpRequest` object, or some other object with conventional properties? The former seems to be the best choice, but I suppose this makes the solution less flexible for integrators.

This seems interesting, but it probably isn't going to happen any time soon. There are several non-trivial higher-priority enhancements ahead of this in the queue, such as support for S3 v4 signatures (#1336), and substantial updates to the documentation. I'd also want to hear a few other users express support for this.

**Comment by joeenzminger** [comment]
I'll start the lobbying process :).  Not urgent for us, either - just something I thought might be a good idea.

**Comment by ajeeshpu** [comment]
:+1:

**Comment by quisse** [comment]
+1

**Comment by rnicholus** [comment]
I don't see this making it into 5.5, as that is currently intended to be a small, focused release. 6.0 is also quite full at the moment (overfull even). Still I can see that it is a desired feature, and I'll keep it in my periphery when planning releases.

**Comment by sedouard** [comment]
+1
